Use `rb_io_wait` function and cache `io` instance.

ioquatix opened the pull request:
https://github.com/github/activerecord-trilogy-adapter/issues/28

matthewd commented on contrib/ruby/ext/trilogy-ruby/cext.c:
```diff
@@ -49,12 +53,29 @@ static size_t trilogy_memsize(const void *ptr) {
+{
+    struct trilogy_ctx *ctx = ptr;
+
+    if (RTEST(ctx->io))
```
Unnecessary?

ioquatix replied:
Mark moveable doesn't check if the field is NULL, IIRC without this it can crash.

matthewd replied:
I was looking at https://github.com/ruby/ruby/blob/1fdaa0666086529b3aae2d509a2e71c4247c3a12/gc.c#L7108 and https://github.com/ruby/ruby/blob/1fdaa0666086529b3aae2d509a2e71c4247c3a12/gc.c#L10395 -- I'm not _100%_ sure it's part of their defined contract, but I imagine it must be true by ossification, even if not originally intended?

Intuitively I just prefer "this slot contains a a possible GC ref", rather than an [even trivial] encoding of "we _might_ need to mark this". 🤷🏻‍♂️

ioquatix replied:
Fixed.

matthewd commented on contrib/ruby/ext/trilogy-ruby/cext.c:
```diff
@@ -183,6 +205,7 @@ static double timeval_to_double(struct timeval tv)
 
 static int _cb_ruby_wait(trilogy_sock_t *sock, trilogy_wait_t wait)
 {
+    struct trilogy_ctx *ctx = sock->user_data;
```
This almost feels like there's a danger of a possible lifetime issue here, but I think it's actually fine.

ioquatix replied:
I didn't look too closely at the lifetimes but I assume the socket should outlive the context?

matthewd commented on contrib/ruby/ext/trilogy-ruby/cext.c:
```diff
@@ -211,10 +234,16 @@ static int _cb_ruby_wait(trilogy_sock_t *sock, trilogy_wait_t wait)
+    }
 
-    return 0;
+    VALUE result = rb_io_wait(ctx->io, RB_INT2NUM(wait_flag), rb_fiber_scheduler_make_timeout(timeout));
```
I acknowledge it's what's already happening on sufficiently-new rubies (along with the IO construction), but `rb_fiber_scheduler_make_timeout` sounds like it might allocate. Is that true? Something we can/should avoid?

ioquatix replied:
There is a small chance I the future it might allocate but for the moment it doesn't, it just embeds a float into the VALUE. I hope at some point in the future to migrate to fixed point timeouts.

matthewd replied:
Bonus question: is this going to affect our min ruby version (which I note we don't seem to currently declare)? I'm guessing this is moderately recent.

ioquatix replied:
No I'll add feature detection, this PR is just what I could manage in the 30 minutes I had before bed time.

eregon replied:
> it just embeds a float into the VALUE

Flonums only work for a subset of all double, if it doesn't fit it will be an allocation.

casperisfine replied:
I think it also always allocate on 32bits?

matthewd reviewed (commented):
Thanks for following up on this @ioquatix!

ioquatix commented:
@matthewd can you please approve and run.

ioquatix commented:
There is one more piece we should consider exposing, which is the "resolve DNS" to use the fiber scheduler hook.

ioquatix commented:
So, writing this PR has made me realise we need a set of shims for the blocking C APIs, e.g. `open`, `addrinfo`, etc. We have the code, it's just not exposed. I supposed we can do that for Ruby 3.3.

We should also extend the test suite to run within a fiber scheduler (probably Async).

matthewd commented:
@ioquatix  is this good to go? It seems fine to me, not sure if you've kept it in draft for a reason, or just not hit the button?

ioquatix commented:
It's probably as good as we can make it.. but I want to review it this weekend. This week has been pretty busy, have not had much time for open source stuff.

casperisfine commented on contrib/ruby/ext/trilogy-ruby/cext.c:
```diff
@@ -211,10 +253,23 @@ static int _cb_ruby_wait(trilogy_sock_t *sock, trilogy_wait_t wait)
+
+#ifdef TRILOGY_RB_IO_WAIT
+    if (ctx->io == Qnil) {
+        ctx->io = rb_io_fdopen(fd, O_RDWR, NULL);
```
`trilogy_data_type` is marked with `RUBY_TYPED_WB_PROTECTED`, so you have to trigger a write barrier here.

Problem is you don't have access to the `VALUE` here.

ioquatix replied:
Okay, so we should pass around the `VALUE` then?

casperisfine replied:
Either that, or create the IO eagerly in a place where you have access to it.

ioquatix replied:
Thanks this is now fixed, I added `user_data` and:

```c
    if (ctx->io == Qnil) {
        VALUE io = rb_io_open_descriptor(rb_cIO, trilogy_sock_fd(sock), FMODE_EXTERNAL, RUBY_Qnil, RUBY_Qnil, NULL);
        RB_OBJ_WRITE(ctx->self, &ctx->io, io);
    }
```

kirs commented:
👋 just wanted to say that  appreciate this WIP PR by Samuel and I'd really really love to see this happen in the upstream, and to be able to use trilogy with fiber scheduler for IO-heavy workloads.

casperisfine commented:
> I'd really really love to see this happen in the upstream, and to be able to use trilogy with fiber scheduler for IO-heavy workloads.

So my knowledge of the fiber scheduler is limited, but AFAIK you can already use it today? My understanding is that this PR make it a bit more efficient, but `rb_wait_for_single_fd` is already enough to be usable.

As for it happening, there's not much work needed, all the blockers are listed in review comments.

ioquatix commented:
I've rebased and updated the PR.

`rb_wait_for_single_fd` is okay but it's a slow path. `rb_io_wait` is the fast path.

@matthewd can you please run CI.

casperisfine commented on contrib/ruby/ext/trilogy-ruby/cext.c:
```diff
@@ -28,26 +38,59 @@ static ID id_socket, id_host, id_port, id_username, id_password, id_found_rows,
     struct trilogy_ctx *ctx = ptr;
-    trilogy_free(&ctx->conn);
+
+    rb_gc_location(ctx->self);
```
```suggestion
    ctx->self = rb_gc_location(ctx->self);
```

casperisfine commented on contrib/ruby/ext/trilogy-ruby/cext.c:
```diff
@@ -28,26 +38,59 @@ static ID id_socket, id_host, id_port, id_username, id_password, id_found_rows,
+    rb_gc_mark_movable(ctx->self);
+
+#ifdef TRILOGY_RB_IO_WAIT
+    if (RTEST(ctx->io))
+        rb_gc_mark_movable(ctx->io);
```
```suggestion
    rb_gc_mark_movable(ctx->io);
```

`gc_mark` don't mind `false/nil`

ioquatix replied:
Thanks I wonder where I got the idea I should check first.

casperisfine commented on contrib/ruby/ext/trilogy-ruby/cext.c:
```diff
@@ -224,6 +274,62 @@ struct rb_trilogy_wait_args {
+        RB_OBJ_WRITE(ctx->self, &ctx->io, rb_io_fdopen(trilogy_sock_fd(sock), O_RDWR, NULL));
+    }
+
+    if (timeout) fprintf(stderr, "_cb_ruby_wait timeout: %ld.%06ld\n", timeout->tv_sec, timeout->tv_usec);
```
```suggestion
```

ioquatix commented:
I recall why I gave up on this PR for a while, we were missing `rb_io_open_descriptor` which is needed to create a "temporary non-owning `FMODE_EXTERNAL` IO instance. Since that's now available in recent CRuby, it is possible to implement it correctly.

ioquatix commented:
I believe all the feedback has been addressed, so this is good to go.

eregon commented on contrib/ruby/ext/trilogy-ruby/cext.c:
```diff
@@ -224,6 +270,66 @@ struct rb_trilogy_wait_args {
+        timeout = NULL;
+    }
+
+    VALUE result = rb_io_wait(ctx->io, RB_INT2NUM(wait_flag), rb_fiber_scheduler_make_timeout(timeout));
```
rb_fiber_scheduler_make_timeout() is not implemented yet on TruffleRuby: https://github.com/search?q=repo%3Aoracle%2Ftruffleruby%20rb_fiber_scheduler_make_timeout&type=code
Yet `#if defined(HAVE_RB_IO_WAIT) && defined(RB_IO_OPEN_DESCRIPTOR) && defined(HAVE_RUBY_FIBER_SCHEDULER_H)` should be true on TruffleRuby.

So could you check for `have_func("rb_fiber_scheduler_make_timeout")` in extconf.rb?

(BTW [that function](https://github.com/ruby/ruby/blob/97e6ad49a4604e7e4ca04da2aaafc63cbd5d29d8/scheduler.c#L270-L278) uses `0.000001f` (32-bit float) shouldn't it be `0.000001` (64-bit double) instead?)

ioquatix replied:
Yes we can fix those issues, thanks for pointing them out.

ioquatix replied:
Here is the fix to prefer double math: https://github.com/ruby/ruby/pull/12803

ioquatix replied:
Okay, fixed.

eregon reviewed (commented):
What's the performance gain with this?
Is it worth the duplication of `_cb_ruby_wait` and additional complexity?

ioquatix commented:
> What's the performance gain with this?

It should be slightly more efficient, but I imagine there isn't much performance difference (except for GC overhead due to the temporary IO objects created).

> Is it worth the duplication of _cb_ruby_wait and additional complexity?

I would say yes. `rb_wait_for_single_fd` creates a temporary IO instance for each call, which creates extra (pointless) garbage. It's better for Ruby extensions that do IO to use a proper IO object IMHO.

ioquatix commented:
I think now that `rb_trilogy_wait_protected` was extracted out, maybe we can simplify this PR a little to address the complexity @eregon mentioned.
